File: products/queryflux/sdlc-ai/observability/libraries/python/observability.py
"""
SDLC Platform Observability Library for Python
Provides integrated logging, metrics, and tracing for Python services.
"""

import asyncio
import json
import time
from contextlib import asynccontextmanager
from contextvars import ContextVar
from typing import Any, Dict, Optional, AsyncGenerator, Callable
from functools import wraps
import logging
import uuid
from datetime import datetime

from prometheus_client import Counter, Histogram, Gauge, start_http_server, REGISTRY
from opentelemetry import trace, baggage, context
from opentelemetry.exporter.jaeger.thrift import JaegerExporter
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor
from opentelemetry.instrumentation.httpx import HTTPXClientInstrumentor
from opentelemetry.propagate import set_global_textmap
from opentelemetry.propagators.b3 import B3MultiFormat
from opentelemetry.sdk.resources import Resource
from opentelemetry.semconv.trace import SpanAttributes

from .structured_logger import (
    StructuredLogger,
    LoggerConfig,
    LogLevel,
    correlation_id_var,
    trace_id_var,
    span_id_var,
    user_id_var,
    tenant_id_var,
    request_id_var,
    session_id_var,
    log_context,
)


class MetricsConfig:
    """Configuration for Prometheus metrics."""

    def __init__(
        self,
        namespace: str = "sdlc",
        subsystem: str = "platform",
        port: int = 8000,
        enable_default_metrics: bool = True,
    ):
        self.namespace = namespace
        self.subsystem = subsystem
        self.port = port
        self.enable_default_metrics = enable_default_metrics


class TracingConfig:
    """Configuration for OpenTelemetry tracing."""

    def __init__(
        self,
        service_name: str = "sdlc-service",
        service_version: str = "1.0.0",
        environment: str = "development",
        enabled: bool = True,
        jaeger_endpoint: str = "http://localhost:14268/api/traces",
        sampling_rate: float = 1.0,
    ):
        self.service_name = service_name
        self.service_version = service_version
        self.environment = environment
        self.enabled = enabled
        self.jaeger_endpoint = jaeger_endpoint
        self.sampling_rate = sampling_rate


class ObservabilityConfig:
    """Configuration for the entire observability stack."""

    def __init__(
        self,
        service_name: str = "sdlc-service",
        service_version: str = "1.0.0",
        environment: str = "development",
        logging_config: Optional[LoggerConfig] = None,
        metrics_config: Optional[MetricsConfig] = None,
        tracing_config: Optional[TracingConfig] = None,
    ):
        self.service_name = service_name
        self.service_version = service_version
        self.environment = environment

        self.logging_config = logging_config or LoggerConfig(
            service=service_name, version=service_version, environment=environment
        )

        self.metrics_config = metrics_config or MetricsConfig()

        self.tracing_config = tracing_config or TracingConfig(
            service_name=service_name,
            service_version=service_version,
            environment=environment,
        )


class BusinessMetrics:
    """Predefined business metrics for SDLC platform."""

    def __init__(self, namespace: str, subsystem: str):
        self.namespace = namespace
        self.subsystem = subsystem

        # HTTP metrics
        self.requests_total = Counter(
            f"{namespace}_{subsystem}_http_requests_total",
            "Total number of HTTP requests",
            ["method", "endpoint", "status_code", "tenant_id", "user_id"],
        )

        self.request_duration = Histogram(
            f"{namespace}_{subsystem}_http_request_duration_seconds",
            "HTTP request duration in seconds",
            ["method", "endpoint", "tenant_id"],
            buckets=[0.1, 0.25, 0.5, 1.0, 2.5, 5.0, 10.0],
        )

        # Error metrics
        self.errors_total = Counter(
            f"{namespace}_{subsystem}_errors_total",
            "Total number of errors",
            ["error_type", "component", "tenant_id"],
        )

        # Authentication metrics
        self.auth_attempts = Counter(
            f"{namespace}_{subsystem}_auth_attempts_total",
            "Total number of authentication attempts",
            ["auth_type", "result", "tenant_id"],
        )

        # Document metrics
        self.documents_processed = Counter(
            f"{namespace}_{subsystem}_documents_processed_total",
            "Total number of documents processed",
            ["operation", "document_type", "tenant_id"],
        )

        self.document_processing_time = Histogram(
            f"{namespace}_{subsystem}_document_processing_duration_seconds",
            "Document processing duration in seconds",
            ["document_type", "operation", "tenant_id"],
            buckets=[0.5, 1.0, 2.5, 5.0, 10.0, 30.0, 60.0],
        )

        # RAG metrics
        self.rag_queries = Counter(
            f"{namespace}_{subsystem}_rag_queries_total",
            "Total number of RAG queries",
            ["query_type", "result_type", "tenant_id"],
        )

        self.rag_query_time = Histogram(
            f"{namespace}_{subsystem}_rag_query_duration_seconds",
            "RAG query duration in seconds",
            ["query_type", "tenant_id"],
            buckets=[0.1, 0.5, 1.0, 2.5, 5.0, 10.0],
        )

        # Vector search metrics
        self.vector_searches = Counter(
            f"{namespace}_{subsystem}_vector_searches_total",
            "Total number of vector searches",
            ["search_type", "index_name", "tenant_id"],
        )

        self.vector_search_time = Histogram(
            f"{namespace}_{subsystem}_vector_search_duration_seconds",
            "Vector search duration in seconds",
            ["search_type", "index_name", "tenant_id"],
            buckets=[0.01, 0.05, 0.1, 0.25, 0.5, 1.0, 2.5],
        )

        # Tenant metrics
        self.active_tenants = Gauge(
            f"{namespace}_{subsystem}_active_tenants",
            "Number of active tenants",
            ["tenant_type"],
        )

        self.tenant_storage = Gauge(
            f"{namespace}_{subsystem}_tenant_storage_bytes",
            "Storage usage per tenant in bytes",
            ["tenant_id", "storage_type"],
        )


class ObservabilityManager:
    """Main observability manager that coordinates logging, metrics, and tracing."""

    def __init__(self, config: ObservabilityConfig):
        self.config = config
        self.logger = StructuredLogger(config.logging_config)
        self.business_metrics = BusinessMetrics(
            config.metrics_config.namespace, config.metrics_config.subsystem
        )
        self.tracer = None
        self._setup_tracing()
        self._setup_metrics()

    def _setup_tracing(self):
        """Setup OpenTelemetry tracing."""
        if not self.config.tracing_config.enabled:
            return

        # Create resource
        resource = Resource.create(
            {
                "service.name": self.config.tracing_config.service_name,
                "service.version": self.config.tracing_config.service_version,
                "deployment.environment": self.config.tracing_config.environment,
            }
        )

        # Create tracer provider
        tracer_provider = TracerProvider(resource=resource)

        # Create and register Jaeger exporter
        jaeger_exporter = JaegerExporter(
            endpoint=self.config.tracing_config.jaeger_endpoint,
            collector_endpoint=self.config.tracing_config.jaeger_endpoint,
        )

        # Add span processor
        span_processor = BatchSpanProcessor(jaeger_exporter)
        tracer_provider.add_span_processor(span_processor)

        # Set global tracer provider
        trace.set_tracer_provider(tracer_provider)

        # Set global propagator
        set_global_textmap(B3MultiFormat())

        # Get tracer
        self.tracer = trace.get_tracer(__name__)

    def _setup_metrics(self):
        """Setup Prometheus metrics."""
        if self.config.metrics_config.port:
            start_http_server(self.config.metrics_config.port)

    def get_logger(self) -> StructuredLogger:
        """Get the structured logger."""
        return self.logger

    def get_tracer(self) -> trace.Tracer:
        """Get the OpenTelemetry tracer."""
        return self.tracer

    def get_business_metrics(self) -> BusinessMetrics:
        """Get business metrics."""
        return self.business_metrics

    @asynccontextmanager
    async def trace_span(
        self,
        name: str,
        attributes: Optional[Dict[str, Any]] = None,
        kind: trace.SpanKind = trace.SpanKind.INTERNAL,
    ) -> AsyncGenerator[trace.Span, None]:
        """Create a traced span with automatic logging and metrics."""
        if not self.tracer:
            yield None
            return

        with self.tracer.start_as_current_span(name, kind=kind) as span:
            if attributes:
                for key, value in attributes.items():
                    span.set_attribute(key, str(value))

            # Log span start
            current_context = context.get_current()
            correlation_id = correlation_id_var.get(current_context)

            self.logger.with_fields(
                span_name=name, span_kind=kind.name, **attributes
            ).info(f"Started span: {name}")

            try:
                yield span
            except Exception as e:
                # Record error in span
                span.record_exception(e)
                span.set_status(trace.Status(trace.StatusCode.ERROR, str(e)))

                # Log error
                self.logger.error(
                    f"Error in span {name}: {str(e)}", error=e, span_name=name
                )

                # Record error metric
                self.business_metrics.errors_total.labels(
                    error_type=type(e).__name__,
                    component=name,
                    tenant_id=tenant_id_var.get(current_context) or "",
                ).inc()

                raise
            else:
                span.set_status(trace.Status(trace.StatusCode.OK))

                # Log span completion
                self.logger.with_fields(span_name=name).info(f"Completed span: {name}")


class ObservabilityMiddleware:
    """ASGI middleware for automatic observability integration."""

    def __init__(self, manager: ObservabilityManager, app):
        self.manager = manager
        self.app = app

    async def __call__(self, scope, receive, send):
        if scope["type"] != "http":
            await self.app(scope, receive, send)
            return

        start_time = time.time()

        # Extract correlation headers
        headers = dict(scope.get("headers", []))

        correlation_id = headers.get(b"x-correlation-id", b"").decode()
        if not correlation_id:
            correlation_id = str(uuid.uuid4())

        trace_id = headers.get(b"x-trace-id", b"").decode()
        span_id = headers.get(b"x-span-id", b"").decode()
        user_id = headers.get(b"x-user-id", b"").decode()
        tenant_id = headers.get(b"x-tenant-id", b"").decode()
        request_id = headers.get(b"x-request-id", b"").decode()
        session_id = headers.get(b"x-session-id", b"").decode()

        # Set correlation data in context
        async with log_context(
            correlation_id=correlation_id,
            trace_id=trace_id,
            span_id=span_id,
            user_id=user_id,
            tenant_id=tenant_id,
            request_id=request_id,
            session_id=session_id,
        ):
            # Start tracing span
            async with self.manager.trace_span(
                f"HTTP {scope['method']} {scope['path']}",
                {
                    "http.method": scope["method"],
                    "http.url": scope["path"],
                    "http.scheme": scope.get("scheme", "http"),
                    "http.user_agent": headers.get(b"user-agent", b"").decode(),
                    "http.remote_addr": scope.get("client", ("", ""))[0],
                    "correlation_id": correlation_id,
                    "tenant_id": tenant_id,
                    "user_id": user_id,
                },
                kind=trace.SpanKind.SERVER,
            ) as span:
                # Wrap send to capture status code
                status_code = 200

                async def wrapped_send(message):
                    nonlocal status_code
                    if message["type"] == "http.response.start":
                        status_code = message["status"]

                        # Add correlation headers to response
                        response_headers = message.get("headers", [])
                        response_headers.extend(
                            [
                                (b"x-correlation-id", correlation_id.encode()),
                            ]
                        )
                        if trace_id:
                            response_headers.append((b"x-trace-id", trace_id.encode()))
                        if span_id and span:
                            response_headers.append(
                                (
                                    b"x-span-id",
                                    span.get_span_context()
                                    .span_id.to_bytes(8, "big")
                                    .hex()
                                    .encode(),
                                )
                            )

                        message["headers"] = response_headers

                    await send(message)

                # Log request start
                self.manager.logger.info(
                    f"HTTP request started: {scope['method']} {scope['path']}",
                    method=scope["method"],
                    path=scope["path"],
                    query_string=scope.get("query_string", b"").decode(),
                    user_agent=headers.get(b"user-agent", b"").decode(),
                )

                try:
                    await self.app(scope, receive, wrapped_send)
                except Exception as e:
                    status_code = 500
                    raise
                finally:
                    # Calculate duration
                    duration = time.time() - start_time

                    # Record metrics
                    self.manager.business_metrics.requests_total.labels(
                        method=scope["method"],
                        endpoint=scope["path"],
                        status_code=str(status_code),
                        tenant_id=tenant_id or "",
                        user_id=user_id or "",
                    ).inc()

                    self.manager.business_metrics.request_duration.labels(
                        method=scope["method"],
                        endpoint=scope["path"],
                        tenant_id=tenant_id or "",
                    ).observe(duration)

                    # Add tracing attributes
                    if span:
                        span.set_attribute("http.status_code", status_code)
                        span.set_attribute("http.duration_ms", duration * 1000)

                    # Log request completion
                    log_level = LogLevel.INFO
                    if status_code >= 400:
                        log_level = LogLevel.WARN
                    if status_code >= 500:
                        log_level = LogLevel.ERROR

                    getattr(self.manager.logger, log_level.value.lower())(
                        f"HTTP request completed: {scope['method']} {scope['path']} {status_code}",
                        method=scope["method"],
                        path=scope["path"],
                        status_code=status_code,
                        duration=duration,
                    )


def instrument_fastapi(app, manager: ObservabilityManager):
    """Instrument FastAPI application with observability."""
    # Add observability middleware
    app.add_middleware(ObservabilityMiddleware, manager=manager)

    # Add OpenTelemetry FastAPI instrumentation
    if manager.tracer:
        FastAPIInstrumentor.instrument_app(app)


def instrument_httpx():
    """Instrument HTTPX client with automatic tracing."""
    HTTPXClientInstrumentor().instrument()


def trace_async(name: str, manager: Optional[ObservabilityManager] = None):
    """Decorator for tracing async functions."""

    def decorator(func: Callable):
        @wraps(func)
        async def wrapper(*args, **kwargs):
            obs_manager = manager or get_global_manager()
            if not obs_manager:
                return await func(*args, **kwargs)

            async with obs_manager.trace_span(name):
                return await func(*args, **kwargs)

        return wrapper

    return decorator


def trace_sync(name: str, manager: Optional[ObservabilityManager] = None):
    """Decorator for tracing sync functions."""

    def decorator(func: Callable):
        @wraps(func)
        def wrapper(*args, **kwargs):
            obs_manager = manager or get_global_manager()
            if not obs_manager:
                return func(*args, **kwargs)

            current_context = context.get_current()

            if obs_manager.tracer:
                with obs_manager.tracer.start_as_current_span(name):
                    return func(*args, **kwargs)
            else:
                return func(*args, **kwargs)

        return wrapper

    return decorator


# Global observability manager instance
_global_manager: Optional[ObservabilityManager] = None


def init_global_observability(config: ObservabilityConfig) -> ObservabilityManager:
    """Initialize global observability manager."""
    global _global_manager
    _global_manager = ObservabilityManager(config)
    return _global_manager


def get_global_manager() -> Optional[ObservabilityManager]:
    """Get global observability manager."""
    return _global_manager


def get_logger() -> StructuredLogger:
    """Get global logger."""
    manager = get_global_manager()
    if manager:
        return manager.get_logger()
    return StructuredLogger(LoggerConfig.default())


def get_tracer() -> Optional[trace.Tracer]:
    """Get global tracer."""
    manager = get_global_manager()
    if manager:
        return manager.get_tracer()
    return None


def get_business_metrics() -> Optional[BusinessMetrics]:
    """Get global business metrics."""
    manager = get_global_manager()
    if manager:
        return manager.get_business_metrics()
    return None

File: products/sdlc-cc/platform/observability/libraries/go/observability.go
package observability

import (
	"context"
	"fmt"
	"net/http"
	"time"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promhttp"
	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/trace"

	"github.com/sdlc-platform/observability/logging"
	"github.com/sdlc-platform/observability/logging/shared"
	"github.com/sdlc-platform/observability/metrics/prometheus"
	"github.com/sdlc-platform/observability/tracing/opentelemetry"
)

// ObservabilityManager combines logging, metrics, and tracing
type ObservabilityManager struct {
	logger           *logging.StructuredLogger
	metricsCollector *prometheus.MetricsCollector
	tracerManager    *opentelemetry.TracerManager
	businessMetrics  *prometheus.BusinessMetrics
	config           *Config
}

// Config holds configuration for the observability stack
type Config struct {
	// Logging configuration
	Logging *logging.LoggerConfig `json:"logging" yaml:"logging"`

	// Metrics configuration
	Metrics *prometheus.Config `json:"metrics" yaml:"metrics"`

	// Tracing configuration
	Tracing *opentelemetry.TracerConfig `json:"tracing" yaml:"tracing"`

	// Service information
	ServiceName    string `json:"service_name" yaml:"service_name"`
	ServiceVersion string `json:"service_version" yaml:"service_version"`
	Environment    string `json:"environment" yaml:"environment"`
}

// DefaultConfig returns default observability configuration
func DefaultConfig() *Config {
	return &Config{
		Logging:        logging.DefaultLoggerConfig(),
		Metrics:        prometheus.DefaultConfig(),
		Tracing:        opentelemetry.DefaultTracerConfig(),
		ServiceName:    "sdlc-service",
		ServiceVersion: "1.0.0",
		Environment:    "development",
	}
}

// NewObservabilityManager creates a new observability manager
func NewObservabilityManager(config *Config) (*ObservabilityManager, error) {
	if config == nil {
		config = DefaultConfig()
	}

	// Override service configuration in sub-configs
	config.Logging.Service = config.ServiceName
	config.Logging.Version = config.ServiceVersion
	config.Logging.Environment = config.Environment
	config.Metrics.Namespace = "sdlc"
	config.Metrics.Subsystem = "platform"
	config.Tracing.ServiceName = config.ServiceName
	config.Tracing.ServiceVersion = config.ServiceVersion
	config.Tracing.Environment = config.Environment

	// Initialize logging
	logger, err := logging.NewStructuredLogger(config.Logging)
	if err != nil {
		return nil, fmt.Errorf("failed to initialize logging: %w", err)
	}

	// Initialize metrics
	metricsCollector := prometheus.NewMetricsCollector(config.Metrics)
	businessMetrics := prometheus.NewBusinessMetrics(metricsCollector)

	// Initialize tracing
	tracerManager, err := opentelemetry.NewTracerManager(config.Tracing)
	if err != nil {
		return nil, fmt.Errorf("failed to initialize tracing: %w", err)
	}

	return &ObservabilityManager{
		logger:           logger,
		metricsCollector: metricsCollector,
		tracerManager:    tracerManager,
		businessMetrics:  businessMetrics,
		config:           config,
	}, nil
}

// GetLogger returns the structured logger
func (om *ObservabilityManager) GetLogger() *logging.StructuredLogger {
	return om.logger
}

// GetMetricsCollector returns the metrics collector
func (om *ObservabilityManager) GetMetricsCollector() *prometheus.MetricsCollector {
	return om.metricsCollector
}

// GetBusinessMetrics returns the business metrics
func (om *ObservabilityManager) GetBusinessMetrics() *prometheus.BusinessMetrics {
	return om.businessMetrics
}

// GetTracerManager returns the tracer manager
func (om *ObservabilityManager) GetTracerManager() *opentelemetry.TracerManager {
	return om.tracerManager
}

// StartMetricsServer starts the metrics HTTP server
func (om *ObservabilityManager) StartMetricsServer() error {
	return om.metricsCollector.StartServer(om.config.Metrics)
}

// Shutdown gracefully shuts down the observability stack
func (om *ObservabilityManager) Shutdown(ctx context.Context) error {
	var errors []error

	// Shutdown tracing
	if err := om.tracerManager.Shutdown(ctx); err != nil {
		errors = append(errors, fmt.Errorf("failed to shutdown tracing: %w", err))
	}

	// Shutdown metrics server
	if err := om.metricsCollector.StopServer(ctx); err != nil {
		errors = append(errors, fmt.Errorf("failed to shutdown metrics server: %w", err))
	}

	if len(errors) > 0 {
		return fmt.Errorf("shutdown errors: %v", errors)
	}

	return nil
}

// ObservabilityMiddleware provides HTTP middleware for observability
type ObservabilityMiddleware struct {
	manager        *ObservabilityManager
	businessHelper *opentelemetry.BusinessTraceHelper
}

// NewObservabilityMiddleware creates new observability middleware
func NewObservabilityMiddleware(manager *ObservabilityManager) *ObservabilityMiddleware {
	return &ObservabilityMiddleware{
		manager:        manager,
		businessHelper: opentelemetry.NewBusinessTraceHelper(manager.tracerManager),
	}
}

// Middleware returns an HTTP middleware function
func (om *ObservabilityMiddleware) Middleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		start := time.Now()

		// Extract correlation data from headers and set in context
		correlationData := shared.ExtractFromHTTP(r)
		ctx := shared.SetCorrelationData(r.Context(), correlationData)
		r = r.WithContext(ctx)

		// Add correlation headers to response
		shared.InjectIntoHTTP(correlationData, w.Header())

		// Start tracing span
		ctx, span := om.manager.tracerManager.StartSpanWithAttributes(
			ctx,
			fmt.Sprintf("HTTP %s %s", r.Method, r.URL.Path),
			[]attribute.KeyValue{
				attribute.String("http.method", r.Method),
				attribute.String("http.url", r.URL.String()),
				attribute.String("http.user_agent", r.UserAgent()),
				attribute.String("http.remote_addr", r.RemoteAddr),
			},
		)
		defer span.End()

		// Wrap response writer to capture status code and response size
		wrapped := &responseWriter{ResponseWriter: w, statusCode: 200}

		// Log request start
		om.manager.logger.Info(ctx, "HTTP request started",
			map[string]interface{}{
				"method": r.Method,
				"path":   r.URL.Path,
				"query":  r.URL.RawQuery,
				"ua":     r.UserAgent(),
			},
		)

		// Process request
		next.ServeHTTP(wrapped, r.WithContext(ctx))

		// Calculate duration
		duration := time.Since(start)

		// Record metrics
		labels := prometheus.Labels{
			"method":   r.Method,
			"endpoint": r.URL.Path,
			"status":   fmt.Sprintf("%d", wrapped.statusCode),
		}

		// Add correlation labels if available
		if correlationData.TenantID != "" {
			labels["tenant_id"] = correlationData.TenantID
		}
		if correlationData.UserID != "" {
			labels["user_id"] = correlationData.UserID
		}
		if correlationData.CorrelationID != "" {
			labels["correlation_id"] = correlationData.CorrelationID
		}
		if correlationData.TraceID != "" {
			labels["trace_id"] = correlationData.TraceID
		}

		om.manager.businessMetrics.RequestsTotal.With(labels).Inc()
		om.manager.businessMetrics.RequestDuration.With(labels).Observe(duration.Seconds())

		// Log request completion
		logLevel := logging.LevelInfo
		if wrapped.statusCode >= 400 {
			logLevel = logging.LevelWarn
			if wrapped.statusCode >= 500 {
				logLevel = logging.LevelError
			}
		}

		om.manager.logger.WithFields(map[string]interface{}{
			"method":      r.Method,
			"path":        r.URL.Path,
			"status_code": wrapped.statusCode,
			"duration":    duration,
		}).Log(ctx, logLevel, "HTTP request completed")

		// Add tracing attributes
		om.manager.tracerManager.SetAttributes(ctx,
			attribute.Int("http.status_code", wrapped.statusCode),
			attribute.Float64("http.duration_ms", duration.Seconds()*1000),
		)

		// Set span status based on status code
		if wrapped.statusCode >= 500 {
			om.manager.tracerManager.SetStatus(ctx, codes.Error, "Server error")
		} else if wrapped.statusCode >= 400 {
			om.manager.tracerManager.SetStatus(ctx, codes.Error, "Client error")
		} else {
			om.manager.tracerManager.SetStatus(ctx, codes.Ok, "Request successful")
		}
	})
}

// responseWriter wraps http.ResponseWriter to capture status code and response size
type responseWriter struct {
	http.ResponseWriter
	statusCode int
	size       int
}

func (rw *responseWriter) WriteHeader(code int) {
	rw.statusCode = code
	rw.ResponseWriter.WriteHeader(code)
}

func (rw *responseWriter) Write(b []byte) (int, error) {
	size, err := rw.ResponseWriter.Write(b)
	rw.size += size
	return size, err
}

// DatabaseHelper provides observability helpers for database operations
type DatabaseHelper struct {
	manager        *ObservabilityManager
	businessHelper *opentelemetry.BusinessTraceHelper
}

// NewDatabaseHelper creates a new database helper
func NewDatabaseHelper(manager *ObservabilityManager) *DatabaseHelper {
	return &DatabaseHelper{
		manager:        manager,
		businessHelper: opentelemetry.NewBusinessTraceHelper(manager.tracerManager),
	}
}

// TraceOperation traces a database operation
func (dh *DatabaseHelper) TraceOperation(
	ctx context.Context,
	operation, table string,
	fn func(context.Context) error,
) error {
	return dh.businessHelper.TraceDatabaseOperation(ctx, operation, table, fn)
}

// TraceOperationResult traces a database operation that returns a result
func TraceOperationResult[T any](
	ctx context.Context,
	dh *DatabaseHelper,
	operation, table string,
	fn func(context.Context) (T, error),
) (T, error) {
	return opentelemetry.WithSpanResult(ctx, dh.manager.tracerManager,
		fmt.Sprintf("DB %s %s", operation, table),
		func(ctx context.Context) (T, error) {
			return fn(ctx)
		},
		attribute.String("db.operation", operation),
		attribute.String("db.table", table),
	)
}

// ErrorHandler provides observability helpers for error handling
type ErrorHandler struct {
	manager *ObservabilityManager
}

// NewErrorHandler creates a new error handler
func NewErrorHandler(manager *ObservabilityManager) *ErrorHandler {
	return &ErrorHandler{
		manager: manager,
	}
}

// HandleError logs and records an error with observability data
func (eh *ErrorHandler) HandleError(ctx context.Context, err error, context map[string]interface{}) {
	// Record error in metrics
	eh.manager.businessMetrics.ErrorsTotal.With(prometheus.Labels{
		"error_type": fmt.Sprintf("%T", err),
		"component":  context["component"].(string),
	}).Inc()

	// Log error with structured data
	fields := make(map[string]interface{})
	for k, v := range context {
		fields[k] = v
	}
	fields["error"] = err

	eh.manager.logger.Error(ctx, "Error occurred", fields...)

	// Record error in tracing
	eh.manager.tracerManager.RecordError(ctx, err,
		attribute.String("error.component", context["component"].(string)),
		attribute.String("error.type", fmt.Sprintf("%T", err)),
	)
}

// Global observability manager instance
var globalObservabilityManager *ObservabilityManager

// InitGlobalObservability initializes the global observability manager
func InitGlobalObservability(config *Config) error {
	var err error
	globalObservabilityManager, err = NewObservabilityManager(config)
	return err
}

// GetGlobalObservability returns the global observability manager
func GetGlobalObservability() *ObservabilityManager {
	if globalObservabilityManager == nil {
		globalObservabilityManager, _ = NewObservabilityManager(DefaultConfig())
	}
	return globalObservabilityManager
}

// Convenience functions using global observability manager
func GlobalLogger() *logging.StructuredLogger {
	return GetGlobalObservability().GetLogger()
}

func GlobalMetricsCollector() *prometheus.MetricsCollector {
	return GetGlobalObservability().GetMetricsCollector()
}

func GlobalBusinessMetrics() *prometheus.BusinessMetrics {
	return GetGlobalObservability().GetBusinessMetrics()
}

func GlobalTracerManager() *opentelemetry.TracerManager {
	return GetGlobalObservability().GetTracerManager()
}
